sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/ReadFromKafkaDoFn.java

```diff
@@ -338,6 +355,9 @@ public ProcessContinuation processElement(
         // When there are no records available for the current TopicPartition, self-checkpoint
         // and move to process the next element.
         if (rawRecords.isEmpty()) {
+          if (expectedOffset > endOffset) {
```

Comment:
Note that `OffsetRange` is `[from, to)` so shouldn't we be checking that `expectedOffset >= endOffset`.

Also, I believe we have this line here because you want to cover the case where you want to read to offset `X` and the cluster only has messages to `X-1` and you don't want to wait for the message at offset `X` to be published. In this case it makes sense to have this if down below within the `for (record : ...) { ... }` loop right after the `outputWithTimestamp`. It would be good to add a comment to this effect as well otherwise others will not understand why we aren't relying on the `tryClaim` to return false.

Reply:
Thanks for the comment, after putting more thought to this, I think it does not make sense to check for the scenario at all since we are getting the endOffset from Kafka Client itself using stopReadTime (will throw exception if no such offset be found). So the waiting scenario would never happen. Just removed the checking section.